src/odfesqlodbc/es_communication.cpp

```diff
@@ -438,9 +472,64 @@ int ESCommunication::ExecDirect(const char* query, const char* fetch_size_) {
         return -1;
     }
     m_result_queue.push(std::unique_ptr< ESResult >(result));
+    if (!result->cursor.empty()) {
+        // If response has cursor, this thread will retrives more results pages asynchronously.
+        auto send_cursor_queries = std::async(std::launch::async, [&]() {
+            SendCursorQueries(result->cursor.c_str());
+        });
+    }
     return 0;
 }
 
+void ESCommunication::SendCursorQueries(const char* _cursor) {
+    if (_cursor == NULL) {
+        return;
+    }
+    try {
+        std::string cursor(_cursor);
+        while (!cursor.empty() && (m_result_queue.size() < m_result_queue_capacity)) {
+            std::shared_ptr< Aws::Http::HttpResponse > response = IssueRequest(
+                SQL_ENDPOINT_FORMAT_JDBC, Aws::Http::HttpMethod::HTTP_POST,
+                ctype, "", "", cursor);
+            if (response == nullptr) {
+                m_error_message =
+                    "Failed to receive response from cursor. "
+                    "Received NULL response.";
+                LogMsg(ES_ERROR, m_error_message.c_str());
+                return;
+            }
+            std::unique_ptr< ESResult > es_result = std::make_unique< ESResult >();
+            AwsHttpResponseToString(response, es_result->result_json);
+            PrepareCursorResult(*es_result);
+            if (es_result->es_result_doc.has("cursor")) {
+                cursor = es_result->es_result_doc["cursor"].as_string();
+                es_result->cursor =
+                    es_result->es_result_doc["cursor"].as_string();
+            } else {
+                SendCloseCursorRequest(cursor);
+                cursor.clear();
+            }
+            m_result_queue.push(std::move(es_result));
+        }
+    } catch (std::runtime_error& e) {
+        m_error_message =
+            "Received runtime exception: " + std::string(e.what());
+        LogMsg(ES_ERROR, m_error_message.c_str());
+    }
+}
+
+void ESCommunication::SendCloseCursorRequest(const std::string& cursor) {
+    std::shared_ptr< Aws::Http::HttpResponse > response =
+        IssueRequest(SQL_ENDPOINT_CLOSE_CURSOR,
+                     Aws::Http::HttpMethod::HTTP_POST, ctype, "", "", cursor);
+    if (response == nullptr) {
```

Comment:
Is it possible to get a null here? Wouldn't you just get an error back and the response would be non-null?

Reply:
MakeRequest can return a null response. Tests in aws sdk has check for it.

ref: https://github.com/aws/aws-sdk-cpp/blob/0a0dddc261903b70354ac74372826b310e43111d/aws-cpp-sdk-core-tests/http/HttpClientTest.cpp